pkg/agent/nodeportlocal/k8s/k8s_controller.go

```diff
@@ -0,0 +1,182 @@
+// +build !windows
+
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"fmt"
+	"hash/fnv"
+	"time"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/portcache"
+
+	corev1 "k8s.io/api/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+const (
+	minRetryDelay = 2 * time.Second
+	maxRetryDelay = 120 * time.Second
+	numQueues     = 4
+)
+
+func bkt(key string, numWorkers uint32) uint32 {
+	h := fnv.New32a()
+	h.Write([]byte(key))
+	bkt := h.Sum32() & (numWorkers - 1)
+	return bkt
+}
+
+type Controller struct {
+	portTable   *portcache.PortTable
+	kubeClient  clientset.Interface
+	queue       []workqueue.RateLimitingInterface
+	Ctrl        cache.Controller
+	CacheStore  cache.Store
+	OldObjStore cache.Store
+}
+
+func NewNPLController(kubeClient clientset.Interface, pt *portcache.PortTable) *Controller {
+	nplCtrl := Controller{
+		kubeClient:  kubeClient,
+		portTable:   pt,
+		OldObjStore: cache.NewStore(cache.MetaNamespaceKeyFunc),
+	}
+	nplCtrl.queue = make([]workqueue.RateLimitingInterface, numQueues)
+	for i := 0; i < numQueues; i++ {
+		nplCtrl.queue[i] = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "nodeportlocal")
+	}
```

Comment:
why multiple queues, typically there is a single queue but multiple worker threads pulling from the same queue?

cc @tnqn

Reply:
Wanted to make sure that the same key is always processed by the same worker. For example, let's say we get a Pod add event and then Pod update event. I think it would be better if both of these events are processed by same worker. Events for different Pods can be processed by different workers.

Reply:
I think the requirement is a specific pod is not processed by multiple workers concurrently and we don't need it's always processed by same worker at different time as worker is stateless, right?

The workqueue has guaranteed the above requirement, please check its comments:
https://godoc.org/k8s.io/client-go/util/workqueue
> * Fair: items processed in the order in which they are added.
> * Stingy: a single item will not be processed multiple times concurrently,
    and if an item is added multiple times before it can be processed, it
    will only be processed once.
> * Multiple consumers and producers. In particular, it is allowed for an
    item to be reenqueued while it is being processed.
> * Shutdown notifications.

Reply:
The workqueue's implementation maintains "processing" and "dirty" items, and won't emit an item until worker marks it as "Done", and if an item is enqueued another time during worker processing it, it will be marked as "dirty" and will be emitted another time after "Done" is called for this item. So the above requirement is guaranteed.

Reply:
Thanks @tnqn. I'll update.

Reply:
Let me know if the latest changes look fine.